[Preprocess #3] Refactor existing Airflow implementation

tskir opened the pull request:
Closes https://github.com/opentargets/issues/issues/3104.

> **Note**
> This is **ready for review,** but please **do not merge.** The base branch is not main, but the previous PR in the series, [Preprocess #2](https://github.com/opentargets/genetics_etl_python/pull/164). This is done so that only changes which are related to this specific PR are displayed. I'll handle merging myself once this is approved.

---

This PR refactors the existing Airflow implementation and separates all code into two conceptually dinstall_dependenciesifferent files:
* `src/airflow/dags/common_airflow.py` contains all boilerplate/initialisation code which can be shared by several DAGs;
* `src/airflow/dags/dag_genetics_etl_gcp.py` is essentially the unchanged existing implementation for the entire workflow populated from `dag.yaml`. However, now that it's using the facilities provided by the common module, it's hopefully more succinct and readable.

You will also notice that I removed the `install_dependencies` step. This is because it calls `install_dependencies_on_cluster.sh`, the same script which is already provided as a cluster initialisation script. So my understanding that this step was completely redundant; however, if I'm mistaken and there was actually more to it, please let me know :upside_down_face:

ireneisdoomed reviewed (commented):
Thanks! These are good additions. General comments:
- I would decouple the creation of the cluster from the installation of the dependencies. Why we decided to do this in the first place was to accelerate development. It very frequently happens that we try running a step, the step fails for whatever reason, and we have to delete/recreate the cluster again. Cluster creation takes a lot of time, updating the project is <1m.
- I still haven't tested this implementation. Is the expectation to have a `dag_genetics_etl_local.py`?

mbdebian commented on src/airflow/dags/common_airflow.py:
I'm not sure I'd put this file in the DAGs folder, Apache Airflow will be interpreting this file in a continuous loop, as it tries to find out changes in the DAGs in the folder.
This _common code_ to all the graphs is more about general configuration than functionality, except for the pyspark job submission helper.
In this sense, I would refactor the configuration into the DAG file, and the helpers would need to be implemented as _Operators_ via the _Providers_ facilities in Apache Airflow.
The only requirement is that all workers have the needed dependencies installed.

d0choa replied:
This is described as a limitation of this approach in this post: 
https://docs.astronomer.io/learn/dynamically-generating-dags#single-file-methods

tskir replied:
We discussed that in person on the 12th of October, but just to confirm @d0choa @mbdebian we are fine with leaving these changes to future iterations, right?

tskir replied:
@d0choa @mbdebian To progress and merge this PR, I'm assuming the above is fine, please let me know if you have any further concerns

tskir commented on src/airflow/dags/common_airflow.py:
```diff
@@ -0,0 +1,135 @@
+    cluster_generator_config = ClusterGenerator(
+        project_id=project_id,
+        zone=zone,
+        master_machine_type=master_machine_type,
```
Actually I think there's an even better way: rather than trying to decide which machine type/cluster size to use for the job, we can configure autoscaling (I've been experimenting with it), and make the cloud worry about it. I think this way is more in line with the Hadoop guidelines.

In any case, would that be all right to leave this change for the next iterations?

ireneisdoomed replied:
Sure

tskir commented on src/airflow/dags/common_airflow.py:
```diff
@@ -0,0 +1,135 @@
+
+
+# Input/output file configuration.
+version = "XX.XX"
```
You're absolutely right, this particular section shouldn't be in this PR, and in a subsequent PR it'll be also implemented differently. Now fixed.

tskir commented on src/airflow/dags/dag_genetics_etl_gcp.py:
```diff
@@ -0,0 +1,74 @@
+    submit_pyspark_job,
+)
+
+SOURCE_CONFIG_FILE_PATH = Path(__file__).parent / "configs" / "dag.yaml"
```
To be more precise, Airflow isn't able to look at files outside the directories we mounted for it in Docker. In this case, it's:
* `dags`
* `configs`
* `logs`
* `plugins`

So because both `dags` and `configs` are mounted, it can access the latter as well.

However, for all files completely outside of the mounted directories, it wouldn't be able to see them.

ireneisdoomed replied:
Thank you for the explanation

tskir commented:
@ireneisdoomed Thank you for your review! I've now rebased against the downstream branch and addressed the comments. I'll address your general questions in two subsequent comments for better readability

tskir commented:
### Cluster creation and initialisation

> I would decouple the creation of the cluster from the installation of the dependencies. Why we decided to do this in the first place was to accelerate development. It very frequently happens that we try running a step, the step fails for whatever reason, and we have to delete/recreate the cluster again. Cluster creation takes a lot of time, updating the project is <1m.

I absolutely wouldn't mind doing this, however, I don't see how this speeds things up. Whether we have:
```
(1) create_cluster >> install_dependencies >> STEP >> delete_cluster
```

Or:
```
(2) create_cluster [and initialise with dependencies] >> STEP >> delete_cluster
```

In each of those cases, if the step fails, the cluster will be deleted and will have to be recreated.

On the other hand, if the cluster is not deleted (for example, because the other steps are still running, or because the automatic deletion has been disabled for development purposes), a task can still be restarted from Airflow UI and will be submitted to the same cluster without recreating it. This is, again, true in both those cases.

It's quite possible that I'm just not getting something here, so happy to chat about it!

tskir commented:
### Ways of running DAG

> I still haven't tested this implementation. Is the expectation to have a dag_genetics_etl_local.py?

Sure, if we want a local DAG, we could implement one — but to be honest I don't see why we would need it. Given all of our data is already in Google Cloud, it probably makes sense to continue using that for all processing, be it development or production. But again, happy to discuss this.

ireneisdoomed commented:
> ### Cluster creation and initialisation
> > I would decouple the creation of the cluster from the installation of the dependencies. Why we decided to do this in the first place was to accelerate development. It very frequently happens that we try running a step, the step fails for whatever reason, and we have to delete/recreate the cluster again. Cluster creation takes a lot of time, updating the project is <1m.
> 
> I absolutely wouldn't mind doing this, however, I don't see how this speeds things up. Whether we have:
> 
> ```
> (1) create_cluster >> install_dependencies >> STEP >> delete_cluster
> ```
> 
> Or:
> 
> ```
> (2) create_cluster [and initialise with dependencies] >> STEP >> delete_cluster
> ```
> 
> In each of those cases, if the step fails, the cluster will be deleted and will have to be recreated.
> 
> On the other hand, if the cluster is not deleted (for example, because the other steps are still running, or because the automatic deletion has been disabled for development purposes), a task can still be restarted from Airflow UI and will be submitted to the same cluster without recreating it. This is, again, true in both those cases.
> 
> It's quite possible that I'm just not getting something here, so happy to chat about it!

Thank you for your explanation. Indeed, when the pipeline is fully deployed and our DAG operates periodically, any failures in a step would trigger a complete rerun, regardless of whether the package installation is linked with cluster creation.

However, for the small period of time where I was testing Airflow to run our steps during development, I frequently found it handy to trigger the steps manually post the cluster creation. 
Imagine the entire DAG has executed and errors pop up in the V2G tasks group. Here's what I was doing:
1. I apply fixes locally and rebuild the project
2. I cleared all tasks except `delete_cluster` so that it has an end state and therefore is not automatically triggered
3. Manually trigger everything downstream `create_cluster` - cluster is created, latest version of the project is installed and `step` starts to run
6. Step 2 again, now the step runs
7. Cluster is deleted

Admittedly this solution is rather hacky, and not the way Airflow is expected to be used. But my main takeaway is that if we're going to use Airflow also during the development phase, it’s important to have an approach that accelerates the workflow.

ireneisdoomed commented:
> ### Ways of running DAG
> > I still haven't tested this implementation. Is the expectation to have a dag_genetics_etl_local.py?
> 
> Sure, if we want a local DAG, we could implement one — but to be honest I don't see why we would need it. Given all of our data is already in Google Cloud, it probably makes sense to continue using that for all processing, be it development or production. But again, happy to discuss this.

I agree with you. My comment was based on the name of the file. I wondered if `dag_genetics_etl_gcp.py` implied setting up a DAG in a context other than GCP.

tskir commented:
### Cluster creation and initialisation
@ireneisdoomed Thank you for your explanations! Especially this part:

> I frequently found it handy to trigger the steps manually post the cluster creation.

Yes, I see how the approach you describe would be difficult to manage. This is how I did it during my DAG testing (https://github.com/opentargets/genetics_etl_python/pull/167), which should be simpler:
1. As soon as the DAG starts running, mark `delete_cluster` as "success" so that it does not trigger automatically. I find this perfectly acceptable in a development setting (but not for future production runs, of course).
2. If a task fails: fix it; rebuild the project; clear the state of the failed task only. It is then restarted automatically and seamlessly on the same cluster.

But this approach depends on another important change, which I think it is time to discuss: in my implementation, a single Dataproc cluster was used for all jobs rather an individual cluster per job.

I think having a single cluster for all jobs is a better approach because of the following:
1. DAG becomes simpler and easier to manage, including resubmitting a failed job. Instead of a three-step task group, each task is, well, a task.
2. There's less overhead compute time spent on cluster initialisation, as it is only done once.
3. If we enable autoscaling (I already experimented with it and it works great), we don't have to worry *at all* about how many resources each job requires. The cluster will have exactly as many workers as are actually required at any given moment.
4. An autoscaling cluster can also make use of secondary preemptible workers, which are approximately 50% cheaper than regular workers.

Combination of points 2, 3, and 4 above make this approach significantly cheaper. But I think equally as importantly, this approach is more in spirit of the cloud technologies. If we have one single-node cluster per job and try to guess how large it should be, we are essentially trying to use it as a VM. In contrast, an autoscaling cluster is a managed solution and it lets the cloud worry how many workers should there be, leaving us to focus on the business logic.

I didn't implement the approach to use a single cluster in this PR because I didn't want to change too many things at any given time. But now that we discussed it, I think it's important to implement this change as well, so that the entire thing is more consistent.

tskir commented:
### Ways of running DAG

> My comment was based on the name of the file. I wondered if `dag_genetics_etl_gcp.py` implied setting up a DAG in a context other than GCP.

@ireneisdoomed Ah, I see what you mean now! Yes, it does makes sense to rename that to just `dag_genetics_etl.py`.

tskir commented:
@ireneisdoomed I addressed the review comments and refactored the way cluster is created and used. Now resubmitting failed tasks should be a piece of cake :upside_down_face: Please let me know if you have any other comments

tskir commented:
@ireneisdoomed I also tested that the DAG has no errors and can be imported by Airflow. This is how it looks like now:
![image](https://github.com/opentargets/genetics_etl_python/assets/10669118/ce562db4-0c52-43cc-96f7-aa3f1b139cb4)

tskir commented:
@ireneisdoomed I've additionally simplified task names, and this is how the DAG looks now:

![image](https://github.com/opentargets/genetics_etl_python/assets/10669118/e1f27dc5-30cb-4c43-9058-c9a5cbc00dd8)

ireneisdoomed commented:
@tskir Ty again for the comments. Based on our discussion:
- The implementation with a single cluster sounds like a good idea. The main arguments for having the steps separated were:
1. Parallelisation. Something that didn't occur with Dataproc Workflows was that independent steps wouldn't run in parallel. Airflow brings parallelisation.
2. Allowing for different cluster configurations based on the steps to be run. One of our requirements that we weren't solving at the moment was the fact that steps could have explicit requirements in terms of the needed infrastructure. I don't know how autoscaling works, but this is something that would be solved by this solution. I think a good exercise is testing if the LD matrix step, the most computationally greedy, would work.

- It is probably a good idea to decouple cluster creation from cluster set up in a following PR once we start working using Airflow. In my experience, installing the dependencies sending the Pig job is 50 seconds vs 5 min of cluster creation. It’s the way I develop at the moment and so far it’s the most agile workflow I’ve had.
